### PR TITLE
Fetch prebuilt binaries securely

### PR DESCRIPTION
**Pull request by p120ph37**
Issue #6

### CODE REVIEW

**Comment by dkrutsko** [comment]
Hey there, so I invested a few hours into trying to solve this problem but unfortunately the fix is not as simple as this. Even though GitHub pages support HTTPS, you can't use custom domains with them. I'd like to use a custom domain because if GitHub ever stops working for us, we could point `node.getrobot.net` to another server and all previous versions of robot-js will still work.

I've also looked at other options such as checking a hash value of the downloaded binaries but that doesn't work because if the binary files get changed (for some emergency) then the hash will no longer match and there would be no way of changing that. I know it's not the best practice to hotfix binaries but in an extreme case, it's very useful.

The only alternative is to host it on a CDN, but I'm not sure if there are any good CDN's that would be willing to host these binaries for free. I'm still hoping that GitHub will one day let us attach certificates to our custom domains but I'm not sure what the status of that is.

**Comment by p120ph37** [comment]
I am aware of the github-pages/https issue (my own blog is a victim of this).
As far as your other concern about being able to survive a github failure, that seems a bit moot since your `package.json` already links to the github domain for source-download.  So if github goes offline, the javascript part of your npm package will fail to download anyway.

**Comment by dkrutsko** [comment]
The JavaScript part is hosted on [NPM](https://www.npmjs.com/package/robot-js). The install script just downloads the node binary from `node.getrobot.net` which happens to be hosted on GitHub. Unless you found something I didn't notice, which would be bad.

**Comment by p120ph37** [comment]
Hmmm, yeah I guess you're right that npm actually does host a copy of the full source (e.g.:  https://registry.npmjs.org/robot-js/-/robot-js-1.0.2.tgz ).

The tricky thing for doing hashes is that you'd need to pre-build all of your binaries before publishing your npm package so that you could include the checksums inside the npm-hosted package.  It would mean that you'd not be able to add additional binary builds for an already-released npm package version.

The problem with hosing on `node.getrobot.net`, as you mentioned, is that HTTPS is not fully-functionaly for gh-pages with custom domains.  But I imagine you want to stay on gh-pages for now because of price (free) and ease of publishing new content (`git push origin`).  Unfortunately, if you are really worried about MitM, even using a non-custom github pages domain wouldn't be sufficient to authenticate the download, since a man in the middle could alter the DNS response (github does not use DNSSEC) and send the request to some other malicious github.io page instead (still within the realm of the same `*.github.io` cert, so no error would be detected).

So without hosting somewhere that you can apply your own TLS cert, the only viable option remaining would be to use public-key signing to validate packages (like Debian does).  Fortunately, Node does have support in its core "crypto" package for this sort of thing... (`crypto.Sign`, `crypto.Verify`)  I'll see if I can set up something reasonable based on that.

**Comment by dkrutsko** [comment]
Yes exactly. And even worse with NPM is that you can't unpublish or override a release because everytime you publish a release, you increase the internal version number. And there are many restrictions with that as well. I mean they have a good reason for doing it but unfortunately it makes project management harder. Which is why robot-js has versions 1.0.1 and 1.0.2 :-P

With regards to this issue, I feel that I might not be taking this as seriously as I should? Maybe I should reach out to some CDN providers and see if I could get something set up.

**Comment by p120ph37** [comment]
Okay, so the latest commit here will use the Crypto module to verify binaries via signature files.  I've pre-populated the public key used for verification with the first of your two GitHub SSH keys ( https://api.github.com/users/dkrutsko/keys ).  If you'd prefer to use your other key, or to create a completely new keypair for signing, the public key value in `Install.js` can be fixed before merging this branch.

The expectation is that for every binary in `robot-js-binaries`, you should also publish a signature file created using your private key.  You can see an example of this in my fork of `robot-js-binaries` (using my own private key, of course) .  If the signature file is missing from the server, or if the signature validation fails, the install script will display a MitM warning.

You can create individual signature files using the openssl command-line tool like this:
```
openssl dgst -sha256 -sign ~/.ssh/github_rsa -out darwin-x64-14.node.sha256 darwin-x64-14.node
```

Or generate signatures for all not-yet-signed binaries like this:
```
find addon -name *.node -exec \[ \! -e {}.sha256 \] \; -exec openssl dgst -sha256 -sign ~/.ssh/github_rsa -out {}.sha256 {} \;
```

You can reformat ssh public key files into PKCS8 format like this:
```
ssh-keygen -f ~/.ssh/github_rsa.pub -e -m pkcs8
```

If for some reason you needed to rebuild a binary file that had already been signed, you'd simply update the signature based on the new binary contents, and as long as you are still using the same private key, the new signature will be accepted by `Install.js`.

**Comment by dkrutsko** [comment]
Wow, I didn't even think of that, very nice solution!!

**Comment by dkrutsko** [comment]
Alright, so I've set up a node script which allows me to generate signatures and SHA-256 hashes with a click of a button. The idea is that I will place them in the [`build_info.json`](https://github.com/Robot/robot-js-binaries/blob/gh-pages/addon/1.0.2/build_info.json) file which will be read by the install script for verification. I have generated a new RSA-4096 key for the signing process. I will make the necessary modifications to the install script under PR #22.

I did have a question though, do you think the modification to the `package.json` file is needed? Specifically the `config.verify` variable?

**Comment by p120ph37** [comment]
The change to `package.json` (the `config.verify` var) seemed a reasonable location to allow user-override of the hash verification.  I think the open question is whether a simple user-bypass of the hash verification is necessary in the first place.  The importance of this depends on how you anticipate the deploy system to be used in the future.

I was taking some inspiration from the way certain Linux distros (e.g. Debian) handle package signing (they generally provide a user-bypass option that is disabled by default), however, Linux distros tend to involve a large collection of different packages signed by the same builder's key.  In that use-case, it often happens that some third party will build or rebuild a package outside of the official build context (where the official signature would be applied), and then want to install from their prebuilt artifact.  This might occur because someone is making a custom-patched version of a package, or building a package which doesn't exist in the mainline distro.  In many of these cases, the builder of this special package will not want to go through the trouble of generating their own keypair, signing their build, and distributing and installing their public key onto the systems where they want to install the package, so they just take the shortcut of disabling the singature-checking, relying instead on their own direct control over the build process and the transport to their deployment-target machines as sufficient to ensure security.

So in other words, since this is a single-project key, and since there isn't really any expectation that others will try to prebuild and host their own binaries (which would be further complicated by needing to supply a download URL anyway, since it's not inferred from the NPM origin), maybe there is no real need to make that setting simple to access (though I'd still recommend keeping it at least as a constant at the top of the install script so that developers can play with it if they want).

**Comment by dkrutsko** [comment]
That's interesting, I guess I'm just having a hard time seeing something like this work in the JavaScript community. Everyone just downloads the required packages and no one really modifies the packages they downloaded. If they did, they might as well update the install script. Maybe if it was an option as part of the root `package.json` file but I don't think that's a very good idea. Regardless, I'll probably leave it just in case.

I also thought a bit more about the signature thing. Looking at how Node does it, they actually include a [SHASUMS256](https://nodejs.org/dist/v6.10.2/SHASUMS256.txt.asc) file as part of their release. But having that whole PGP thing is difficult since Node doesn't have native support for it. I'm thinking since there's only one key (Node has [multiple](https://github.com/nodejs/node#release-team)), we can have one file, maybe called `signatures.txt`, that contains the signed values for all the files (including `build_info.json`). It doesn't need to be signed itself but all the important files would be signed.

**Comment by p120ph37** [comment]
The nice thing about the "config" key is that it doesn't need to be an option of the root `package.json` in order for `npm config set robot-js:verify false`, issued in the root context, to affect the install of the `robot-js` dependency package.

NPM also includes signatures in their release `package.json` files, but of course that ties the binaries to the `package.json`, which was something we were avoiding here.  For Node, yeah, they are using GPG, as is Debian - the effect is the same, but the tooling is a little less nice for automation inside Node itself.  Using the OpenSSL signing mechanism allows us to leverage Node's core Crypto module directly.

And yes, having one master signature file (perhaps formatted as JSON for easy parsability) could work fine.  Probably would want one signature file per release version if that's the way you do it.  The big projects like Node and Debian have more complex requirements because of the large number of authorized build users.

And yes, the signature file does not need to be signed or encrypted itself, since the fact that the signature matches the binary and the public key is itself all that matters - the file is just a convenient way to find the signature you want to test against the binary+key.

**Comment by dkrutsko** [comment]
Thanks for clearing that up, and teaching me something new :-)

I'll work on getting this implemented when I get home today.

**Comment by dkrutsko** [comment]
@p120ph37: okay, I finally found the time I needed to get this wrapped up. PR #22 has been updated and will be merged soon. If you're interested and you have the time, please take a look at the new install script and see if there is anything wrong. It seems to be working and the implementation looks solid but you never know with these things :-/

Thanks again for all your help!

**Comment by p120ph37** [comment]
Checked your diff, looks reasonable.  I probably would have used json for the signatures file, but your text parsing looks solid, so that works too. 👍

**Comment by dkrutsko** [comment]
Thanks for taking a look!! I took inspiration from the shasums files I found in Node releases, the parsing is also way simpler for languages that don't support JSON out of the box.
